### PR TITLE
fix(period): introduce helper to calculate relative period start dates

### PR DESCRIPTION
**Pull request by HendrikThePendric**
We introduced support for relative periods for standard reports recently in #138. Afterwards I realised that, when we are dealing with fixed periods, we get a period-start-date and we append that to the URL and also to the GET query parameters. For relative periods we were not doing that.

I thought that this might be linked to the problem outlined in [this JIRA issue-comment](https://jira.dhis2.org/browse/DHIS2-8844?focusedCommentId=35722&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-35722). I also consulted the documentation and [here](https://docs.dhis2.org/2.34/en/developer/html/dhis2_developer_manual_full.html#webapi_viewing_analytical_resource_representations) it says **we do need to** provide a date field when working with relative periods. So even if it turns out that the problem described in the JIRA issue is unrelated, this fix is needed anyway.

I have implemented a helper that calculates start dates for relative periods and am now also including this start date in the URL and the GET requests. The function that calculates the start dates has test coverage, and I double-checked the correctness of the outcomes by comparing the values with values I could produce in the data-visualisation app.

To test the functionality you need to have a view a "Standard report" of type "HTML" which is configured to use relative dates. In the Sierra Leone database we find a few of these standard reports of type HTML, but some have issues. I would advise this:
- Use the report called "Immunization: Overview (HTML-based with SVG)" in the Standard reports section
- First edit the report and change some settings:
    - Make sure to check at least one relative period, for example "yesterday" (Not "Reporting day", that will let you chose a fixed period)
    - Check "Reporting Period" under "Reporting parameters"
    - Uncheck "Organisation Unit" under "Reporting parameters"
- After saving the report, click on it again to view it
- In the dialog that pops up, chose a relative date and click generate. Now you can inspect the report.

To verify things are working, you can check various aspects:
- There is a date query parameter in the browser URL 
- There is a date field in the GET request URL
- And most importantly, when you check the iframe content, or GET request response, you see that the `dhis2.report.date` variable is populated with that date (see below)

Screenshot 2020-10-01 at 14 32 42

### CODE REVIEW

File: src/utils/periods/__tests__/relativePeriods.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import { getRelativePeriodStartDate } from '../relativePeriods.js'
+
+describe('getRelativePeriodStartDate', () => {
+    const originalDateNow = Date.now
+
+    beforeAll(() => {
+        global.Date.now = jest.fn(() =>
```
Comment (ismay) [inline src/utils/periods/__tests__/relativePeriods.js:7]: Jest has a nice way of mocking globals with [spyOn](https://jestjs.io/docs/en/jest-object#jestspyonobject-methodname), it requires a little less manual effort. For example:

```js
jest.spyOn(global.Date, 'now').mockImplementation(() => 'someDate')
```

Afterwards you can clean it up with [mockRestore](https://jestjs.io/docs/en/mock-function-api#mockfnmockrestore). (I'm not sure if the jest version you have here supports it though)

File: src/utils/periods/relativePeriods.js [hunk 2]
```diff
@@ -1,179 +1,224 @@
 import i18n from '@dhis2/d2-i18n'
 import { fixedPeriodTranslations } from './fixedPeriods'
 
+const RELATIVE_PERIOD_CODES = {
```
Comment (ismay) [inline src/utils/periods/relativePeriods.js:4]: Any reason you went with an object here instead of string constants? I think string constants would give more certainty in that you can be sure that they won't be overwritten. Which feels appropriate for this usecase.

File: src/utils/periods/relativePeriods.js [hunk 3]
```diff
@@ -195,3 +240,118 @@ export const flattenedRelativePeriods = RELATIVE_PERIODS.reduce(
 
 export const isFixedPeriodType = periodType =>
     !!fixedPeriodTranslations[periodType]
+
+const formatYyyyMmDd = timestamp => {
```
Comment (ismay) [inline src/utils/periods/relativePeriods.js:244]: I was wondering, if moment's available here, maybe that would be easier than using the date api for all the date based calculations in this file?

File: src/utils/periods/relativePeriods.js [hunk 4]
```diff
@@ -195,3 +240,118 @@ export const flattenedRelativePeriods = RELATIVE_PERIODS.reduce(
...
+    const mm = (date.getMonth() + 1).toString().padStart(2, '0')
+    const dd = date
+        .getDate()
+        .toString()
+        .padStart(2, '0')
+
+    return `${yyyy}-${mm}-${dd}`
+}
+
+const getCurrentDate = () => new Date(Date.now())
```
Comment (ismay) [inline src/utils/periods/relativePeriods.js:256]: Is this not the same as just calling `new Date()`?

File: src/utils/periods/relativePeriods.js [hunk 5]
```diff
@@ -195,3 +240,118 @@ export const flattenedRelativePeriods = RELATIVE_PERIODS.reduce(
...
+
+const getPreviousMonday = date => {
+    const day = date.getDay()
+    const mondayTimeStamp = date.getDate() - day + (day === 0 ? -6 : 1)
```
Comment (ismay) [inline src/utils/periods/relativePeriods.js:260]: This is a bit hard to parse.

File: src/utils/periods/relativePeriods.js [hunk 6]
```diff
@@ -195,3 +240,118 @@ export const flattenedRelativePeriods = RELATIVE_PERIODS.reduce(
...
+}
+
+export const getRelativePeriodStartDate = periodCode => {
+    const startDateCalutator = startDateCalculators[periodCode]
+
+    if (!startDateCalutator) {
+        return null
+    }
+
+    return formatYyyyMmDd(startDateCalutator())
```
Comment (ismay) [inline src/utils/periods/relativePeriods.js:356]: Personally I think I'd go with a simpler name for this function. The pascalcase formatted date reads a bit awkward.

**Review by ismay — COMMENTED** [review]
Worked as expected when tested with debug/dev 👍. I asked a couple of questions for things that weren't clear to me. I do have to say that I'm not super familiar with this particular app, so I might miss more intricate details here.

File: src/utils/periods/__tests__/relativePeriods.js (hunk 1 above)
Comment (HendrikThePendric) [reply src/utils/periods/__tests__/relativePeriods.js]: Addressed in 3ecb1d7

File: src/utils/periods/relativePeriods.js (hunk 2 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js]: Can't remember why I ended up with those. I gradually moved stuff around and refactored. Perhaps at some point in time these codes were needed in a different file and therefore it made more sense to export them as an object. In the current state of the codebase, your suggestion makes perfect sense though. Implemented in b9329cd

File: src/utils/periods/relativePeriods.js (hunk 3 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:242]: Perhaps this logic is going to be extracted from the reports-app and integrated into [dhis2/analytics ](https://github.com/dhis2/analytics). So I tried to use helpers [as present there](https://github.com/dhis2/analytics/blob/master/src/components/PeriodDimension/utils/fixedPeriods.js#L537-L546). Therefore I'd like to keep it the way it is.

File: src/utils/periods/relativePeriods.js (hunk 4 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:254]: This is makes it easier to mock the current date, since you then only have to mock `Date.now` and not the entire `Date` class/constructor. I was reading about this [here](https://codewithhugo.com/mocking-the-current-date-in-jest-tests/)
> Using `new Date(Date.now())` makes for code that is a lot easier to test. Mocking a function that returns a number (like `Date.now`) is a lot easier than mocking a constructor.

File: src/utils/periods/relativePeriods.js (hunk 5 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:258]: I agree it's hard to digest, but I don't really see a way around this. "Our" weeks start on a Monday but `date.getDay()` will return `0` for Sunday, so a "real/normal" week starts on Sunday. My conclusion is that this is hard to parse, because it's a bit quirky/complex.

File: src/utils/periods/relativePeriods.js (hunk 6 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:354]: This is similar to https://github.com/dhis2/reports-app/pull/192#discussion_r500906665. I'd like to sticvk with the naming applied in `@dhis2/analytics`.

File: src/utils/standardReport/index.js [hunk 7]
```diff
@@ -40,8 +41,14 @@ export const appendOrgUnitsAndReportPeriodToQueryString = (
...
-            const period = parsePeriod(reportPeriod.selectedPeriod)
-            baseStr += `&date=${period.startDate}`
+            const relativePeriodStartDate = getRelativePeriodStartDate(
+                reportPeriod.selectedPeriod
+            )
+            const fixedPeriod =
+                !relativePeriodStartDate &&
+                parseFixedPeriod(reportPeriod.selectedPeriod)
+            const startDate = relativePeriodStartDate || fixedPeriod.startDate
+            baseStr += `&date=${startDate}`
```
[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/utils/standardReport/index.js]: I agree this is pretty nasty. I have introduced a few helper functions in 204ece4 to make this more palatable. I think this is now pretty clear.

File: src/utils/periods/relativePeriods.js (hunk 4 above)
Comment (ismay) [reply src/utils/periods/relativePeriods.js:254]: Ah I see. Maybe a comment explaining that would be helpful. Otherwise it's easy to assume for someone not familiar with the app that it can be factored out (if you don't look at the tests).

File: src/utils/periods/relativePeriods.js (hunk 5 above)
Comment (ismay) [reply src/utils/periods/relativePeriods.js:258]: Yeah makes sense. We could quote your explanation in a comment. That would already help me grasp more quickly what's going on.

File: src/utils/periods/relativePeriods.js (hunk 6 above)
Comment (ismay) [reply src/utils/periods/relativePeriods.js:354]: Sounds good 👍

File: src/utils/periods/relativePeriods.js (hunk 3 above)
Comment (ismay) [reply src/utils/periods/relativePeriods.js:242]: Maybe a small comment explaining that, somewhere in the beginning of this file, would be helpful. Just so people are aware that that's where the code style comes from?

File: src/utils/standardReport/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (ismay) [reply src/utils/standardReport/index.js]: Yeah I agree, that looks a lot easier to grasp to me, nice 👍

**Review by ismay — APPROVED** [review]
Looks good 👍. Left some feedback on potential explanatory comments, but nothing that affects functionality.

File: src/utils/periods/relativePeriods.js (hunk 4 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:258]: Yeah I agree a comment would be useful, added in d97053b0ad8a0916c627156a7591d4b682c15d7e. Having said that, probably it would become evident that this could not be factored out pretty quickly, since the tests would start failing.

File: src/utils/periods/relativePeriods.js (hunk 5 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:266]: Added in eb454bb

File: src/utils/periods/relativePeriods.js (hunk 3 above)
Comment (HendrikThePendric) [reply src/utils/periods/relativePeriods.js:242]: added in bec9d6e
